### PR TITLE
[go_router] [shell_route] Fix BottomNavigationBar might lost current index in long URIs.

### PR DESCRIPTION
**Pull request by omar-hanafy**
I used the shell route example in my app. I noticed that when I navigate to a sub route of a specific BottomNavigationBarItem route e.g "/c/details/, the _calculateSelectedIndex returns the default index -> 0. That's because in if statement we use location == "BottomNavigationBarItem Location", which returns false in all cases if the location =  /c/details. I did change location == "location" to location.startsWith("location"), so the _calculateSelectedIndex determine the selected index based on the first route of the location not the exact one.

before:

after:

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/go_router/example/pubspec.yaml [hunk 1]
```diff
@@ -8,19 +8,19 @@ environment:
   flutter: ">=3.3.0"
 
 dependencies:
-  adaptive_dialog: ^1.2.0
-  adaptive_navigation: ^0.0.4
-  collection: ^1.15.0
-  cupertino_icons: ^1.0.2
+  adaptive_dialog: ^1.8.0+1
```
Comment (chunhtai) [inline packages/go_router/example/pubspec.yaml:11]: can you revert these?

**Review by johnpryan — APPROVED** [review]

**Comment by johnpryan** [comment]
@omar-hanafy it looks like you need to update the pubspec and CHANGELOG, even though this is only changing a sample

**Comment by omar-hanafy** [comment]
@johnpryan I updated the pubspec and CHANGELOG if there is anything left I did not do just let me know.

File: packages/go_router/example/pubspec.yaml [hunk 2]
```diff
@@ -27,4 +27,4 @@ dev_dependencies:
     sdk: flutter
 
 flutter:
-  uses-material-design: true
+  uses-material-design: true
```
Comment (chunhtai) [inline packages/go_router/example/pubspec.yaml:30]: Can you revert this change?

File: packages/go_router/example/lib/shell_route.dart [hunk 3]
```diff
@@ -77,6 +77,24 @@ class ShellRouteExampleApp extends StatelessWidget {
               ),
             ],
           ),
+
+          /// The third screen to display in the bottom navigation bar.
```
Comment (chunhtai) [inline packages/go_router/example/lib/shell_route.dart:81]: Since you are touching the example, can you write a simple test in example/test? just to make sure c/details can display the right content.

**Review by chunhtai — COMMENTED** [review]
I didn't see the changelog and version bump in the diff.

**Comment by omar-hanafy** [comment]
> I didn't see the changelog and version bump in the diff.

I added a change log and diff before, but when i tried to merge there was a conflict and i did not accept my changes.
Should i update the changelog and version bump again?

**Comment by chunhtai** [comment]
yes, please do so

**Review by chunhtai — APPROVED** [review]
LGTM
